src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.ValueDispatchNode.cs

```diff
@@ -0,0 +1,256 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal partial class LocalRewriter
+    {
+        private abstract partial class DecisionDagRewriter
+        {
+            /// <summary>
+            /// A node in a tree representing the form of a generated decision tree for classifying an input value.
+            /// </summary>
+            private abstract class ValueDispatchNode
+            {
+                protected virtual int Height => 1;
+#if DEBUG
+                protected virtual int Weight => 1;
+#endif
+                public readonly SyntaxNode Syntax;
+
+                public ValueDispatchNode(SyntaxNode syntax) => Syntax = syntax;
+
+                /// <summary>
+                /// A node representing the dispatch by value (equality). This corresponds to a classical C switch
+                /// statement, except that it also handles values of type float, double, decimal, and string.
+                /// </summary>
+                internal sealed class SwitchDispatch : ValueDispatchNode
+                {
+                    public readonly ImmutableArray<(ConstantValue value, LabelSymbol label)> Cases;
+                    public readonly LabelSymbol Otherwise;
+                    public SwitchDispatch(SyntaxNode syntax, ImmutableArray<(ConstantValue value, LabelSymbol label)> dispatches, LabelSymbol otherwise) : base(syntax)
+                    {
+                        this.Cases = dispatches;
+                        this.Otherwise = otherwise;
+                    }
+                    public override string ToString() => "[" + string.Join(",", Cases.Select(c => c.value)) + "]";
+                }
+
+                /// <summary>
+                /// A node representing a final destination that requires no further dispatch.
+                /// </summary>
+                internal sealed class LeafDispatchNode : ValueDispatchNode
+                {
+                    public readonly LabelSymbol Label;
+                    public LeafDispatchNode(SyntaxNode syntax, LabelSymbol Label) : base(syntax) => this.Label = Label;
+                    public override string ToString() => "Leaf";
+                }
+
+                /// <summary>
+                /// A node representing a dispatch based on a relational test of the input value by some constant.
```

Comment:
>A [](start = 20, length = 1)

I think this responds to the request for a comment explaining Left/Right/WhenTrue/WhenFalse.